server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java

```diff
@@ -645,6 +645,17 @@ private IndexMetadata(
         assert numberOfShards * routingFactor == routingNumShards : routingNumShards + " must be a multiple of " + numberOfShards;
     }
 
+    public IndexMetadata forDataNodePersistence() {
```

Comment:
Could we avoid this completely if we let `RemoveCorruptedShardDataCommand` take the index UUID and custom path as input? Does not seem completely unreasonable to me to put that burden on the operator for this hopefully rare case.

Reply:
Yeah quite possibly. It already has a `-d` option that lets you specify the index directory, maybe we can just require that if using a custom data path. Technically a breaking change, but only if doing something deeply ill-advised on a deeply unusual setup.

Reply:
I looked further at this and noticed that we need the cluster state to map the index name to its UUID even for indices that aren't using a custom data path. If we got rid of that we'd need to drop the whole "look up index by name" feature from `RemoveCorruptedShardDataCommand`. I think that's too much of a break. In the distant past we required users to specify the path to the shard directory and experience showed this didn't work very well.

We do usually write the index metadata in each index directory too so we could mostly do the same thing by scanning through all the indices on the node, but this is controlled by `gateway.write_dangling_indices_info` so it's not guaranteed.

I think the solution described in https://github.com/elastic/elasticsearch/issues/74149 would help fix this too since it would give data nodes a way to represent info about their local shards independently of the cluster state. Until then I guess we're stuck needing to track at least a little bit of cluster state on each data node.